### PR TITLE
Use abci-tag proc_macro for event attributes

### LINKED ISSUE #214: Proc macro attr for event attributes
## Summary

Would be nice to have a proc macro attr for event attributes -> 
```rust
#[abci_tag(PORT_ID_ATTRIBUTE_KEY)]
pub struct PortIdAttribute {
    pub port_id: PortId,
}
```
The generated code would be -> 
```rust
pub struct PortIdAttribute {
    pub port_id: PortId,
}

impl From<PortIdAttribute> for Tag {
    fn from(attr: PortIdAttribute) -> Self {
        Tag {
            key: PORT_ID_ATTRIBUTE_KEY.parse().unwrap(),
            value: attr.port_id.to_string().parse().unwrap(),
        }
    }
}
```

-> [Original comment on #198](https://github.com/cosmos/ibc-rs/pull/198#discussion_r1012317911)
____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by DaviRain-Su**
Closes: #214 

## Description

This is my implementation of the `abci_tag proco marco` implementation of the feeling that there is still a toy Everyone see what other improvements, this implementation is to solve the problems raised in this issue.

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/abci-tag/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "abci-tag-proc-marco"
```
Comment (plafer) [inline crates/ibc/abci-tag/Cargo.toml:2]:
```suggestion
name = "abci-event-attr-key-proc-macro"
```

File: crates/ibc/abci-tag/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
+use quote::quote;
+
+use proc_macro::TokenStream;
+use proc_macro2::Ident;
+use syn::{parse_macro_input, DataStruct, DeriveInput, Ident as SynIdent};
+
+#[proc_macro_attribute]
+pub fn abci_tag(args: TokenStream, item: TokenStream) -> TokenStream {
```
Comment (plafer) [inline crates/ibc/abci-tag/src/lib.rs:8]:
```suggestion
pub fn abci_event_attr_key(args: TokenStream, item: TokenStream) -> TokenStream {
```

Could please change the name? I find this more descriptive.

File: crates/ibc/abci-tag/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          pub #field_name: #ty,
+        }
+
+        impl From<#name> for tendermint::abci::EventAttribute {
+          fn from(attr: #name) -> Self {
+            EventAttribute {
+              key: #args.parse().unwrap(),
+              value: attr.#field_name.as_str().parse().unwrap(),
+              index: false,
+            }
```
Comment (plafer) [inline crates/ibc/abci-tag/src/lib.rs:59]: This is the old way to do it. Can you have the return value be as it was, e.g. 

```rust
impl From<PortIdAttribute> for abci::EventAttribute {
    fn from(attr: PortIdAttribute) -> Self {
        (PORT_ID_ATTRIBUTE_KEY, attr.port_id.as_str()).into()
    }
}
```

**Review by plafer — CHANGES_REQUESTED** [review]
Thank you for this!

Could you also implement the change for 

- the structs found in `core::ics02_client::events`
- `ModuleEventAttribute` in `crate::events`

**Comment by romac** [comment]
May I ask what problem this is solving and what's the scope for this? Is it only used to derive the From implementation? To me it sounds a bit overkill to introduce a whole new crate and a proc macro to save on 4 lines of code per event that won't change substantially in the future.

**Comment by romac** [comment]
How we really want to save on those 4 lines, how about using a declarative macro instead? eg.

```rust
event_attr!(PortIdAttribute, PORTID_ATTR_KEY, port_id);
```

**Comment by plafer** [comment]
I agree... I even prefer no declarative macro at all since it's just clearer what is being implemented for the struct.

Should we just close this PR and the issue? @romac @hu55a1n1 WDYT?

**Comment by hu55a1n1** [comment]
What I liked about the proc v/s declarative macro approach was that you didn't have to specify anything other than the attr key const, but I agree this is overkill as there are more additions in the PR than there are deletions (`+104 −87`). :+1: 
I think this was a good experiment but I am also in favour of closing the PR and issue.

**Comment by plafer** [comment]
Thank you @DaviRain-Su for this. Unfortunately we'll close it for the reasons discussed.

**Comment by DaviRain-Su** [comment]
> Thank you @DaviRain-Su for this. Unfortunately we'll close it for the reasons discussed.

Understand

**Pull request closed without merging**